packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -1682,6 +1684,27 @@ class FocusManager with DiagnosticableTreeMixin, ChangeNotifier {
   // update.
   final Set<FocusNode> _dirtyNodes = <FocusNode>{};
 
+  // Stores the node that was focused before the app lifecycle changed.
+  // Will be restored as the primary focus once app is resumed.
+  FocusNode? _suspendedNode;
+
+  @override
+  void didChangeAppLifecycleState(AppLifecycleState state) {
+    if (state == AppLifecycleState.resumed) {
+      if (_suspendedNode != null) {
+        assert(_focusDebug(() => 'marking node $_suspendedNode to be focused'));
+        _markedForFocus = _suspendedNode;
+        _suspendedNode = null;
+        applyFocusChangesIfNeeded();
+      }
+    } else if (_primaryFocus != rootScope) {
+      assert(_focusDebug(() => 'suspending $_primaryFocus'));
+      _markedForFocus = rootScope;
+      _suspendedNode = _primaryFocus;
```

Comment:
What should happen if something changes the focus while the app is not in "resumed" mode? (Imagine a timer being set to return focus to something at a later time.) Should setting the primary focus to something other than the `rootScope` modify the `_suspendedNode`, or should it clear it and set the primary focus?

Reply:
That's a really good question. I'm not aware of an app that changes focus on a timer, so there probably isn't a precedent to follow.

But if a developer implements a feature that can change the focus without user input, it probably makes sense to just let it happen, i.e. null out the `_suspendedNode` and set the primary focus. I'll update this function accordingly.

Reply:
Well, it doesn't have to be as simple as a timer, I was just thinking of the simplest way to describe the issue. For a more real-world case, Imagine a slide-in animation that sets the focus on a text field when the animation is done.  If you switch the focus away during the animation, this would also cause the same problem.

Your solution looks good, though.

Reply:
One thing just occurred to me, though: we might not want to expose `didChangeAppLifecycleState` as a public function on the `FocusManager`, since it's not really something that should be called outside of the binding.  Perhaps it would be better aggregate instead of inherit here: create a private class that extends `WidgetsBindingObserver` and let the `FocusManager` have a private instance of that class.

Reply:
done!